### ISSUE TITLE
Clean source code

### ISSUE THREAD

**Issue by project-am** [Ben987/Bondage-College#10]
Iirc, you wrote that you would like to use this project to learn JS. Since I alrady have a decent amount of JS experience, I would like to help you by providing my opinions and maybe even cleaning up some of your source (if that is ok with you), though as I wrote, I won't have a lot of time over the next few weeks.

I have already started becoming a bit familiar with your source code, so I see a few quick-gains in terms of architecture and a few improvements concerning your draw-related code. Before doing all that stuff, though, I want to check if you actually want to have such PRs...

Most of the mentioned fixes are good for mobile data volume, battery life and performance, while they should also greatly improve the code architecture, making the program more generic, hence easier to extend later on (in case you want to add a feature, or make the day repeatable, maybe randomly putting together classes)

## Quick improvements

- [x] cache downloaded files instead of re-downloading them whenever you need them (mobile data volume!)
    - [x] images
    - [x] CSV files
- [ ] switch all scripts to [strict-mode](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Strict_mode)
- [ ] use double-buffering
- [ ] instead of creating functions and variables in the global scope, use a namespace for your application
- [ ] put all functions into objects (by chapter?) and just use `window.app.c003.routines[dynamicCall]();` instead of creating a function by string concatenation and creating a new function object
- [ ] get a linter (for example eslint) with a sane preset (I like the airbnb one)
- [ ] use a build-step to bundle and optimize the code (for example using webpack)
- [ ] use a tree for you interface, in which you register areas, so you can find out the clicked area very quickly in a generic manner (at the moment you run the check on every single thing)
- [ ] remove double-toggles. For example, you have `C002_FirstClass_Classroom_MildredSubdueSuccess` and `C002_FirstClass_Classroom_MildredSubdueFailed`, which might lead to the ambitous situation of both being true. Instead, make an enum-like structure and fill in a number for the status. Alternatively, use Monadic Options (so you can have `Some(true)`, `Some(false)` or `None`) or in the worst case `null`.
- [ ] store actor-info on the actor instead of using separate variables for each chapter
- [ ] (if someone has terrifying performance problems only!) use Numbers to work with actors and items instead of doing string comparisons everywhere
- [ ] use instantiated objects with explicit fields for actors instead of an array
- [ ] make the game a [progressive app](https://developers.google.com/web/progressive-web-apps/)

## Big changes

- [ ] scale canvas with screen-size (I already experiment with that in a local fork ;) )
- [ ] abstract drawing functions away (maybe make a little engine lib you can re-use; I really love your ideas and hope this is not your last game :D)
- [ ] only include the necessary objects (actors) in the images with transparent background, so you can easily re-use and re-arrange them
- [ ] make character-images composable instead of uploading a separate image for every possible combination
- [ ] only re-draw updated areas of the screen and only when really necessary (I sometimes have a flickering screen - it would be remedied)

## Food for thought

- [ ] ES6 has been available for quite some time, offering lots of improvements to JS (Template Strings, Promises, Symbols, Maps, Iterators, Class sugar, etc.)
- [ ] Using WebGL, you might be able to add some interesting visual effects via shaders
- [ ] Using WebAudio, you can set a better atmosphere
- [ ] [Vibration API](https://developer.mozilla.org/en-US/docs/Web/API/Vibration_API) for PWM vibrators (as in smartphone vibrate function, baka), but nevertheless................

**Comment by Ben987** [comment]
Thanks a lot for all the ideas and suggestions, I like it.  There's already a few of them I want to implement but my main problem is time, I'm already playing with transparent background right now.  Scaling the game to the screen sounds like an easy one, do you have code for that?

If I had knew javascript from the start I would have probably done things differently but I had to learn the language first.  I would probably have created objects for everything like chapters and actors.  But since nobody seems to be having performance issues yet I guess arrays are fine.

**Comment by project-am** [comment]
> If I had knew javascript from the start

There is always a first time :) I really love the idea of your game, that's why I want to help you from the programming side of things (I am not an artist, so I cannot help you with the images :/ ). What I can do is tell you how you can improve your code, stabilize it, decouple it, make it easy to work with and implement useful patterns and principles (like SOLID, DRY, KISS, etc.)

> I would probably have created objects for everything like chapters and actors. But since nobody seems to be having performance issues yet I guess arrays are fine.

Objects are mainly useful for you as a developer, not for performance reasons. The fields make it clear which index is used for what. Do the comparison youself:

```js
const Actors = [[
  'Alice',
  182,
  104,
  28,
  44,
  929,
  73,
]];
```

versus

```js
const Actors = [{
  name: 'Alice',
  health: 182,
  mana: 104,
  strength: 28,
  agility: 44,
  speed: 929,
  avatarID: 73,
}];
```

> But since nobody seems to be having performance issues yet

As a matter of fact, I do. There are some scene which are constantly flickering. On a AAA gaming-grade desktop PC. I think, your performance would benefit the most by caching stuff and only redrawing necessary parts of the viewport. I forgot it in the list above, but double-buffering would also get rid of the flickering. However there is still the problem with mobile data volume and energy consumption. On top of that, I have no idea how slow the scenes render on a low-end smartphone (my smartphone has more processing power than my laptop, sry o_O). Well I will add double-buffering to the low-hanging fruit list, as it is fairly simple (use a second hidden canvas to which you draw, then copy the content of the second canvas as bitmap to the first canvas).

> do you have code for that

I tried re-writing the render and game-code, while still using your original data (just to see how far I can get stuff). The resizing of the frame and images is working, but I had to abandon it because of time :D Well, I don't think it makes a whole lot of sense to keep a separate copy alive with all those features if I can help bring them upstream, see it more as a little challenge for me personally.

My code is based on OpenGL viewport principles, which means that I put the origin `[0.0, 0.0]` into the top-left corner. The bottom-right corner has the coordinates `[1.0, 1.0]`. Anything which is rendered can only be placed in between those coordinates. The actual pixels are calculated. I don't feel comfortable uploading the whole thing right now, so here are the most important parts for scaling:

![screenshot](https://i.imgur.com/0XIahSx.png)

```js
/**
 * Init engine
 * Set it up with the original aspect ratio
 */
boco.engine.init = function engineInit() {
    const vpWidth = Math.max(document.documentElement.clientWidth, window.innerWidth || 0);

    // Get the context
    this._canvas = document.querySelector('.js-vp');
    this._ctx = this._canvas.getContext('2d');

    // Set up the _canvas
    this.borderOffset = 2;
    this._canvasWidth = vpWidth - 2 * this.borderOffset;
    this._canvasHeight = vpWidth * (675 / 1200) - 2 * this.borderOffset;// keep original aspect ratio
    this._canvas.setAttribute('width', this._canvasWidth.toString());
    this._canvas.setAttribute('height', this._canvasHeight.toString());

    // Other settings
    this._ctx.font = this._canvasHeight / 30 + 'px sans';
};

/**
 * Calculate sprite region
 * Accepts $p1 or $p2 to not be set, in case image proportions should be kept
 *
 * @param {String} $src Sprite-source
 * @param {Array.<Number>} $p1 2D Point // OPTIONAL
 * @param {Array.<Number>} $p2 2D Point // OPTIONAL
 * @return {Array.<{origin: Array.<Number>, width: Number, height: Number}>}
 */
boco.engine._getSpriteArea = function engine_getSpriteArea($src, $p1, $p2) {
    let height = 0;
    let width = 0;

    if ($p2[0]) {
        width = ($p2[0] - $p1[0]) * this._canvasWidth;
    }

    if ($p2[1]) {
        height = ($p2[1] - $p1[1]) * this._canvasHeight;
    }

    if (!$p2[0]) {
        width = this._sprites[$src].naturalWidth * (height / this._sprites[$src].naturalHeight);
    }

    if (!$p2[1]) {
        height = this._sprites[$src].naturalHeight * (width / this._sprites[$src].naturalWidth);
    }

    return {
        origin: [$p1[0] * this._canvasWidth, $p1[1] * this._canvasHeight],
        width: width,
        height: height,
    };
};

/**
 * Draws a line from the first point to the second one
 *
 * @param {Array.<Number>} $p1 X and Y coordinates of first point
 * @param {Array.<Number>} $p2 X and Y coordinates of second point
 */
boco.engine.drawLine = function engineDrawLine($p1, $p2) {
    this._ctx.beginPath();
    this._ctx.moveTo($p1[0] * this._canvasWidth, $p1[1] * this._canvasHeight);
    this._ctx.lineTo($p2[0] * this._canvasWidth, $p2[1] * this._canvasHeight);
    this._ctx.stroke();
};

/**
 * Draw sprite
 * Accepts $p1 or $p2 to not be set, in case image proportions should be kept
 *
 * @param {String} $src Sprite-source
 * @param {Array.<Number>} $p1 2D Point // OPTIONAL
 * @param {Array.<Number>} $p2 2D Point // OPTIONAL
 * @param {Function} $cb Callback when drawing is done
 */
boco.engine.drawSprite = function engineDrawSprite($src, $p1, $p2, $cb) {
    const draw = () => {
        this._sprites[$src].removeEventListener('load', draw);

        if ($p2) {
            const area = this._getSpriteArea($src, $p1, $p2);
            this._ctx.drawImage(
                this._sprites[$src],
                area.origin[0],
                area.origin[1],
                area.width,
                area.height
            );
        }
        else {
            this._ctx.drawImage(
                this._sprites[$src],
                $p1[0] * this._canvasWidth,
                $p1[1] * this._canvasHeight
            );
        }

        $cb && $cb();
    };

    this._getSprite($src).then(draw, () => {});
};

/**
 * Draw text
 *
 * @param {String} $text
 * @param {Array.<Number>} $pos 2D Point
boco.engine.drawText = function engineDrawText($text, $pos) {
    this._ctx.fillStyle = '#333';
    this._ctx.fillText($text, $pos[0] * this._canvasWidth, $pos[1] * this._canvasHeight)
};
```

Sample usage:

```js
boco.gfx.drawIGMenu = function gfxDrawIGMenu() {
    const engine = boco.engine;

    engine.drawLine([0, .9], [1, .9]);
    engine.drawLine([.85, .9], [.85, 1]);
    engine.drawText(boco.state.currentTime.toLocaleTimeString(), [.88, .96]);

    this.drawCharacterButton();
};

boco.engine.init();
boco.gfx.drawIGMenu();
```

**Comment by Ben987** [comment]
Thanks a lot :)  I'm not sure I'll be able to use your code but I'll see if I can make it work without too many problems.  As you can guess I'm more of an artist and a writer than a programmer, most of what you listed is foreign language to me.

I'll also try someday to make an object to use for every character, including the player, an object that will keep all the stats and the functions to use.  So the actors and the player will have the same methods.

**Comment by project-am** [comment]
oh, well, I am sorry. Sometimes I don't even realize that I use terminology which is not commonly known.

I have to say that you really did an amazing job! I really wish to help you make this great and a breeze to extend, because all I can do is dive into code (really not an artist >.< ; tbh, I cannot wait for how you continue the game's story :D ). I hope that I can take away all the complicate stuff without changing too much of what you already created, so just wait for some useful PRs :)

Ah, yeah, and if you need help with a feature or don't understand the jack I tell you, just ask and I will give you a hint (you do want to learn JS, duh). I'll try to check the issues regularly.

**Comment by Ben987** [comment]
Thanks!  I'm glad you like my little game.  I actually tried to do a cache system for the images and it seems to stop most of the flickering.  Basically, once I load an image I keep it in an array that I reuse if I need to load it again.  That way it can only flicker the first time the image is loaded.  Try it out :)

**Comment by project-am** [comment]
I just wanted to turn off my computer when I saw that you uploaded the cache system two minutes prior. So I'm not a creep stalking your commits D:

I thought I should take a look at the new Sarah content (heh, can't control myself >.<) and also took a look at the new caching system. Since it's really small, I opened a PR for a more elegant solution, which makes use of objects (you haven't taken a look at them yet, have you :D).

The flickering is gone and I have a quick and solid experience. Awesome, thank you for implementing it so quickly! :) 🎉 🎉 🎉 🎉
